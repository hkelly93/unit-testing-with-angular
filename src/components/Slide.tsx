import React, { useState } from 'react';
import classnames from 'classnames';
import { ThemeConsumer } from '../context/ThemeContext';
import useEventListener from '@use-it/event-listener';

type Props = {
  title?: string;
  subtext?: React.Component | string;
  children: React.Component | string;
  footer?: React.Component | string;
  animate?: boolean;
  notes?: string;
};

let hasSeenFirstNote = false;

const Slide = ({
  title,
  subtext,
  children,
  footer,
  animate = true,
  notes = '',
}: Props) => {
  const [showNotes, setShowNotes] = useState(false);

  useEventListener('keydown', (event: any) => {
    const { keyCode } = event;
    const isN = keyCode === 78;

    if (notes) setShowNotes(showNotes && isN ? false : true);

    if (isN) {
      hasSeenFirstNote = true;
    }
  });

  const getNoteIcon = (theme: any) => (
    <div className={theme['has-notes']}>
      ðŸ“£ {!hasSeenFirstNote && <span>Press n for notes</span>}
    </div>
  );

  return (
    <ThemeConsumer>
      {(theme: any) => (
        <div className={classnames(theme.slide, { [theme.animate]: animate })}>
          {title && <h1>{title}</h1>}

          {subtext && <div className={theme.subtext}>{subtext}</div>}

          {children}

          {footer ? footer : null}

          {showNotes && <div className={theme.notes}>{notes}</div>}

          {notes && getNoteIcon(theme)}
        </div>
      )}
    </ThemeConsumer>
  );
};

export default Slide;
