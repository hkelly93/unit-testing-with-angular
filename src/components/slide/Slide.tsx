import React from 'react';
import classnames from 'classnames';
import { ThemeConsumer } from '../../context/ThemeContext';

type Props = {
  title?: string;
  subtext?: React.Component | string;
  children: React.Component | string;
  animate?: boolean;
};

export default ({ title, subtext, children, animate = true }: Props) => (
  <ThemeConsumer>
    {(theme: any) => (
      <div className={classnames(theme.slide, { [theme.animate]: animate })}>
        {title && <h1 data-testid="title">{title}</h1>}

        {subtext && (
          <div className={theme.subtext} data-testid="subtext">
            {subtext}
          </div>
        )}

        {children}
      </div>
    )}
  </ThemeConsumer>
);
