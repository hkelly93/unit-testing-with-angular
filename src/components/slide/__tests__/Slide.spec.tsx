import React from 'react';
import { render, cleanup } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';

import ThemeContext from '../../../context/ThemeContext';
import Slide from '../Slide';
import theme from '../../../theme.module.scss';

describe('Slide Component', () => {
  afterEach(cleanup);

  const props = {};

  const getComponent = (p = props, children: React.Component | string) => (
    <ThemeContext.Provider value={theme}>
      <Slide {...p}>{children}</Slide>
    </ThemeContext.Provider>
  );

  describe('Animate', () => {
    test('has the "animate" class when animate is enabled (default)', () => {
      const { container } = render(getComponent({}, 'Test'));
      const { firstChild } = container;

      expect(firstChild).toHaveClass(theme.slide);
      expect(firstChild).toHaveClass(theme.animate);
    });

    test('has the "animate" class when animate is enabled', () => {
      const { container } = render(getComponent({ animate: true }, 'Test'));
      const { firstChild } = container;

      expect(firstChild).toHaveClass(theme.slide);
      expect(firstChild).toHaveClass(theme.animate);
    });

    test('does not have the "animate" class when animate is disabled', () => {
      const { container } = render(getComponent({ animate: false }, 'Test'));
      const { firstChild } = container;

      expect(firstChild).toHaveClass(theme.slide);
      expect(firstChild).not.toHaveClass(theme.animate);
    });
  });

  describe('title', () => {
    test('does not appear if it is not passed', () => {
      const { queryByTestId } = render(getComponent({ animate: true }, 'Test'));

      expect(queryByTestId('title')).toBeNull();
    });

    test('does appear if it is passed', () => {
      const title = 'mytitle';
      const { queryByTestId } = render(
        getComponent({ animate: true, title }, 'Test')
      );

      const titleElement = queryByTestId('title');
      expect(titleElement).toHaveTextContent(title);
      expect(titleElement).toMatchSnapshot(); // Validate the tag.
    });
  });

  describe('subtext', () => {
    test('does not appear if it is not passed', () => {
      const { queryByTestId } = render(getComponent({ animate: true }, 'Test'));

      expect(queryByTestId('subtext')).toBeNull();
    });

    test('does appear if it is passed', () => {
      const subtext = 'mysubtext';
      const { queryByTestId } = render(
        getComponent({ animate: true, subtext }, 'Test')
      );

      const subtextElement = queryByTestId('subtext');
      expect(subtextElement).toHaveTextContent(subtext);
      expect(subtextElement).toHaveClass(theme.subtext);
      expect(subtextElement).toMatchSnapshot(); // Validate the tag.
    });
  });
});
