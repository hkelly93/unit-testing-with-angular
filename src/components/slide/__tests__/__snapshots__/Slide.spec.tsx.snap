// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Slide Component subtext does appear if it is passed 1`] = `
<div
  class="subtext"
  data-testid="subtext"
>
  mysubtext
</div>
`;

exports[`Slide Component title does appear if it is passed 1`] = `
<h1
  data-testid="title"
>
  mytitle
</h1>
`;
